src/element-internals.ts

```diff
@@ -251,7 +251,37 @@ if (!window.CustomStateSet) {
   window.CustomStateSet = CustomStateSet;
 }
 
-if (!window.ElementInternals) {
+export function isElementInternalsSupported(): boolean {
+  if (!window.ElementInternals) {
+    return false;
+  }
+
+  class ElementInternalsFeatureDetection extends HTMLElement {
+    internals: IElementInternals;
+
+    constructor() {
+      super();
+      this.internals = this.attachInternals();
+    }
+  }
+  const randomName = `element-internals-feature-detection-${Math.random().toString(36).replace(/[^a-z]+/g, '')}`
+  customElements.define(randomName, ElementInternalsFeatureDetection);
+  return [
+    "shadowRoot",
+    "form",
+    "states",
+    "willValidate",
+    "validity",
+    "validationMessage",
+    "labels",
+    "setFormValue",
+    "setValidity",
+    "checkValidity",
+    "reportValidity"
+  ].every(prop => prop in new ElementInternalsFeatureDetection().internals)
```

Comment:
Not sure if it's a concern, but the new object is created here for every element in the property name array.

Reply:
Thank you, sorted out in a2a22cd